File: isaac_deploy/isaac_ros_deploy_converters/isaac_ros_deploy_converters/tensor_utils.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Shared utilities for building CUDA-buffer-compatible tensor messages."""

import numpy as np
from tensor_msgs.msg import ExperimentalTensor

# numpy dtype string -> DLPack (code, bits). Lanes are scalar (1).
DTYPE_TO_DLPACK = {
    'float32': (2, 32),
    'float64': (2, 64),
    'int8': (0, 8),
    'int16': (0, 16),
    'int32': (0, 32),
    'int64': (0, 64),
    'uint8': (1, 8),
}


def make_tensor(data: np.ndarray) -> ExperimentalTensor:
    """Create a contiguous ExperimentalTensor message from a numpy array."""
    tensor = ExperimentalTensor()
    data = np.ascontiguousarray(data)
    tensor.shape = list(data.shape)
    dtype_str = str(data.dtype)
    if dtype_str not in DTYPE_TO_DLPACK:
        raise ValueError(
            f"Unsupported numpy dtype '{dtype_str}' for TensorList message. "
            f'Supported dtypes: {list(DTYPE_TO_DLPACK.keys())}'
        )
    tensor.dtype_code, tensor.dtype_bits = DTYPE_TO_DLPACK[dtype_str]
    tensor.dtype_lanes = 1
    # DLPack permits an empty stride vector for contiguous row-major tensors.
    # This also avoids NumPy's noncanonical zero strides on size-one axes.
    tensor.strides = []
    tensor.byte_offset = 0
    tensor.data = data.tobytes()
    return tensor

File: isaac_deploy/isaac_deploy_core/core/tensor_utils.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "isaac_deploy_core/core/tensor_utils.hpp"

#include <algorithm>
#include <unordered_map>

namespace isaac_deploy_core {

  expected < std::vector < int64_t >> compute_reorder_indices(
    const std::vector < std::string > &source_names,
    const std::vector < std::string > &target_names) {
    // Build index map from source names to their positions.
    std::unordered_map < std::string, int64_t > source_index_map;
    for (size_t i = 0; i < source_names.size(); ++i) {
      if (source_index_map.contains(source_names[i])) {
        return tl::unexpected(
          make_error(
            Error::Code::kInvalidArgument,
            "Duplicate name in source: " + source_names[i]));
      }
      source_index_map[source_names[i]] = static_cast < int64_t > (i);
    }

    // Compute indices for target order.
    std::vector < int64_t > indices;
    indices.reserve(target_names.size());
    for (const auto & name : target_names) {
      auto it = source_index_map.find(name);
      if (it == source_index_map.end()) {
        return tl::unexpected(
          make_error(
            Error::Code::kNotFound,
            "Name not found in source: " + name));
      }
      indices.push_back(it->second);
    }

    return indices;
  }

  torch::Tensor reorder_tensor(
    const torch::Tensor & tensor, const std::vector < int64_t > &indices,
    int64_t dim)
  {
    auto index_tensor =
      torch::tensor(indices, torch::TensorOptions().dtype(torch::kLong).device(tensor.device()));
    return tensor.index_select(dim, index_tensor);
  }

  torch::Tensor reorder_tensor(
    const torch::Tensor & tensor, const torch::Tensor & index_tensor,
    int64_t dim)
  {
    return tensor.index_select(dim, index_tensor);
  }

}  // namespace isaac_deploy_core
